R/build_sce.R

```diff
@@ -0,0 +1,113 @@
+#' Build the SCE object from the counts matrix
+#'
+#' @param counts Counts matrix with rownames corresponding to gene names and colnames corresponding to cell barcodes.
+#'   Can be a counts matrix with intron counts specified by -I or an alevin-fry "USA" matrix,
+#'   with intron counts marked by "-U" and ambiguous counts "-A".
+#' @param include_unspliced Whether or not to include the unspliced reads in the counts matrix.
+#'   If TRUE, the main "counts" assay will contain unspliced reads and spliced reads and an additional "spliced"
+#'   assay will contain spliced reads only. If TRUE, requires that data has been aligned to a reference contianing
+#'   spliced and unspliced reads.
+#'   Default is TRUE.
+#' @param round_counts Logical indicating in the count matrix should be rounded to integers on import.
+#'   Default is TRUE.
+#'
+#' @return SingleCellExperiment object containing either just a counts assay with spliced cDNA only if
+#'   `include_unspliced` is FALSE. If `include_unspliced` is TRUE, the counts assay will contain both spliced and unspliced
+#'    counts and the spliced assay will contain the counts for just the spliced cDNA.
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # build a SCE object with only spliced cDNA as the main counts assay
+#' build_sce(counts)
+#'
+#' # build a SCE object with unspliced cDNA as the main counts assay and spliced
+#' # counts as a second assay
+#' build_sce(counts,
+#'           include_unspliced = TRUE)
+#'
+#'}
+build_sce <- function(counts,
+                      include_unspliced = TRUE,
+                      round_counts= TRUE){
+
+
+  if(!is.logical(include_unspliced)){
+    stop("include_unspliced must be set as TRUE or FALSE")
+  }
+
+  if(!is.logical(round_counts)){
+    stop("round_counts must be set as TRUE or FALSE")
+  }
+
+  # define if counts matrix has any unspliced reads
+  has_unspliced <- any(grep("-[IU]$", rownames(counts)))
+
+  if(include_unspliced & !has_unspliced){
+    stop("No counts corresponding to intronic reads detected.
+          If `include_unspliced` is TRUE a reference with spliced and unspliced reads must be used.")
+  }
+
+  # if using intron mode and want to include the unspliced data get counts for both unspliced and spliced
+  if(include_unspliced & has_unspliced) {
+
+    unspliced <- collapse_intron_counts(counts, which_counts = c("unspliced"))
+    spliced <- collapse_intron_counts(counts, which_counts = c("spliced"))
```

Comment:
Realizing that going back in time, we would want to change this `which_counts  =  "total"`, but that could break things. But I think we can do it if we modify `collapse_intron_counts` to allow either option with the same result, that is probably worth doing.

So for consistency, the following suggestion, and the downstream effects thereof, unfortunately.

```suggestion
    total <- collapse_intron_counts(counts, which_counts = "total")
    spliced <- collapse_intron_counts(counts, which_counts = "spliced")
```